Python Backend Refactor

MCLiii opened the pull request.

MCLiii committed:
Watch out for snakes

MCLiii committed:
add Config  file and Database functions

MCLiii committed:
add components directory

MCLiii committed:
Update README.md

jrvollmer commented:
TODOs for refactor (may be incomplete or include things that don't end up getting implemented):
- [x] README
  - [x] Add FastAPI and any other frameworks/libraries that are used or packages that have to be installed to the "Libraries" section (we should probably rename this section to "libraries, frameworks, and runtimes," "software/technology used," or something like that)
  - [x] Update "Running the Dashboard"
  - [x] Update "Contributing to the Dashboard"
  - [x] If we use a dependency manager (see below), add a section about using it to install Python dependencies
- [x] I'm not sure how many Python dependencies we'll have by the end of the refactor, but we should consider using a Python dependency manager (e.g. Pipenv, micropipenv, or pip-tools)
- [x] `Dockerfile` will have to be updated to install all necessary dependencies and also reflect changes for running the dashboard, if any
- [x] If the data format path changes (please not unless it's really out of place with the new backend), we'll have to update our scripts in the front end, our `pull-run` scripts, and `.gitmodules`
- [x] We ~~may~~will have to change the `pull-run` scripts if any backend paths or the commands to run the dashboard are changed
- [x] `DataGenerator/` may have to change
- [x] Update versioning guidelines for Python backend
- [ ] Backend:
  - [x] Graph API
  - [x] Single values API
  - [x] Telemetry (TCP)
  - [ ] Different run configurations
    - [ ] Run with `engineering-data-distributor`
      - [ ] "individual" and "competition" configurations
    - [x] "dev" run configuration
  - [x] Data recording
  - [ ] Remove nodejs Backend code/files

---

- [ ] Stress testing

MCLiii committed:
Create setup.py

MCLiii committed:
Update README.md

MCLiii committed:
Amend versioning.md for python backend

MCLiii committed:
db query now returns pandas df

jrvollmer commented on Backend/setup.py:
```diff
@@ -0,0 +1,13 @@
+    packages=['core', 'framework', 'components'],
+    url='badgerloop.org',
+    license='',
+    author='badgerloop',
```
```suggestion
    url='badgersolarracing.org',
    license='',
    author='Badger Solar Racing Software Team',
```

jrvollmer replied:
A bit nit-picky, but could you move it into [the "Create setup.py" commit](b7095edf292950cc4aec7d5c8fcdaf2c80fdb2a3) so that it never shows up as badgerloop?

jrvollmer commented on Backend/setup.py:
```diff
@@ -0,0 +1,13 @@
+
+setup(
+    name='chase-car-dashboard-backend',
+    version='5.0.0',
```
In the same commit, can you please update the version in `chase-car-dashboard/package.json` to `5.0.0`? Also, future note, we should move the addition of `setup.py`, the updates to `README.md` and `versioning.md`, and the change of versions in `package.json` to their own commit as the most recent commit in the PR

jrvollmer replied:
ty

jrvollmer reviewed (changes requested):
A few minor changes for the time being. I will do an in-depth review as soon as I get the chance. @MCLiii, have you tried running this with Docker? If not, tag the most recent commit (`git tag v5.0.0-alpha`), push the tag (`git push origin v5.0.0-alpha`), and after a few minutes (or whenever the image shows up in the container registry), use the `linux-pull-run.sh` script to pull and run the image in the individual configuration (`./linux-pull-run.sh -c individual -t v5.0.0-alpha`)

MCLiii committed:
Make datagenerator work

MCLiii committed:
ctrl + c now works

jrvollmer commented on README.md:
Not necessary, but would definitely help for future members: Thoughts on adding either a section to this README or a new README in `Backend/` that briefly describes how to add a new API endpoint for the backend? E.g.
```
1. Create new file called <api_endpoint>.py
2. Create router:
\```
from fastapi import APIRouter

router = APIRouter()
\```
3. Define endpoint: <Generic example endpoint>
4. Add to `component_router.py`: <Example code snippet>
```

jrvollmer commented on DataGenerator/src/server.js:
```diff
@@ -102,7 +103,9 @@ const SERVER = NET.createServer((socket) => {
-
+    //write end symbol
+    buf1.write("</bl>",buffOffset)
+    console.log(buf1)
```
```suggestion
```
Let's not log anything unnecessary. Hypocritical coming from me, but I don't want you to have to deal with the same storm of useless console logs that I did

jrvollmer commented on Backend/core/db.py:
```diff
@@ -0,0 +1,48 @@
+
+def insert_data(data_dict):
+    # Get the timestamp from the dictionary
+    timestamp = datetime.utcnow().replace(
+        hour=(data_dict['tstamp_hr'] + 6) % 24,
+        minute=data_dict['tstamp_mn'],
+        second=data_dict['tstamp_sc'],
+        microsecond=data_dict['tstamp_ms']
+    )
```
Note for a potential task that ties in with sc1-driver-io: We should really be getting the timezone/hour dynamically so that we don't run into issues when going to competition or switching to daylight savings.

MCLiii replied:
Yeah we are working on putting in a unix timestamp into the byte array

jrvollmer replied:
Sg. I'll mark as resolved then

jrvollmer commented on Backend/config.py:
This would be a good place to record the different values that you're computing for data analysis. If you add a README to the backend (or add a section about adding new routes/computed data), you should add a note/step about adding new computed values to this file.

jrvollmer replied:
Looks good. README note was just a potential idea, so marking as resolved

jrvollmer commented on Backend/core/comms.py:
```diff
@@ -0,0 +1,113 @@
+    print(byte_length)
+
+def unpack_data(data):
+    #print(solar_car_connection)
```
Let's remove this from the commit in which it was added

jrvollmer commented on Backend/core/comms.py:
```diff
@@ -0,0 +1,113 @@
+                break
+
+            #print("start index:", start_index, "end index:", end_index)
+            #print(solar_car_connection)
```
Can remove unused print statements from the commit(s) in which they were added

jrvollmer commented on Backend/core/comms.py:
```diff
@@ -0,0 +1,113 @@
+        format_string += types[data_format[key][1]]
+        byte_length += data_format[key][0]
+        properties.append(key)
+    print(byte_length)
```
```suggestion
```
Again, unnecessary print

MCLiii committed:
Fix bug in db

jrvollmer commented on Backend/core/comms.py:
```diff
@@ -0,0 +1,113 @@
+def start_comms():
+    gen_format_str(config.DATAFORMAT_PATH)
+    tcp = TCP()
+    tcp.listen_tcp('127.0.0.1', 4003)
```
```suggestion
    tcp.listen_tcp(TCP_ADDRESS, TCP_PORT)
```
Can we make 4003 a constant, as well as any other ports and addresses?

jrvollmer replied:
ty

jrvollmer commented on Backend/core/db.py:
```diff
@@ -0,0 +1,65 @@
+        hour=(data_dict['tstamp_hr'] + 6) % 24,
+        minute=data_dict['tstamp_mn'],
+        second=data_dict['tstamp_sc'],
+        microsecond=data_dict['tstamp_ms']
```
```suggestion
        millisecond=data_dict['tstamp_ms']
```

jrvollmer commented on Backend/core/db.py:
```diff
@@ -0,0 +1,65 @@
+    del data_dict['tstamp_sc']
+    del data_dict['tstamp_ms']
+    # Insert data into Redis time series database
+    # print(timestamp.timestamp())
```
```suggestion
```
:hurtrealbad:

jrvollmer reviewed (changes requested):
@MCLiii, more changes

MCLiii committed:
Update setup.py for name change

Co-authored-by: jrvollmer <72111610+jrvollmer@users.noreply.github.com>

MCLiii committed:
Add latest data query

MCLiii committed:
Add configuration settings

MCLiii committed:
Change package.json to work with python backend

jrvollmer commented on Backend/config.py:
```diff
@@ -0,0 +1,12 @@
+REDIS_PORT = 6379
+REDIS_DB = 0
+
+CAR_IP = '127.0.0.1'
```
```suggestion
CAR_IP = '192.168.1.15'
```

jackwherry committed:
make backend work

jackwherry committed:
frontend changes

MCLiii committed:
initial graph api implementation

jackwherry committed:
use milliseconds, fix tz offset

jackwherry committed:
make Excel files contain readable dates

jackwherry committed:
bump versions

MCLiii committed:
Convert to unixtime and change big endian to little in data generator

MCLiii committed:
add lte driver, add exception handling, change <bl> to <bsr>

MCLiii committed:
fix disconnect bug and fix packet latency

MCLiii committed:
Graph backend

Add decimation algos:
- Avg for analog values
- Min-max for bools
- TODO Specify algo for each value in data format

MCLiii committed:
Update and improve graph frontend

- Use new graph API and adjust graph-related components accordingly
- Reduce graphSelectModal rerender frequency by removing unnecessary
  dependencies from its useMemo()
- Handle fetch errors/timeouts with new graph API
- Add unix timestamps to faux-queue and misc-constants

Co-authored-by: James <jamesvollmer02@gmail.com>

jrvollmer committed:
Ignore Backend/*.egg-info

MCLiii committed:
Async fetch, slow down tcp reconnect, update data format

MCLiii committed:
Explain magic numbers

MCLiii committed:
set up docker

MCLiii committed:
add aiohttp to setup.py

MCLiii committed:
tweaking with docker

MCLiii committed:
Amend pull-run script

MCLiii commented:
@jrvollmer Linux pull run script and dockerfile is updated, I don't have any windows machine, can you change the windows script if possible?

MCLiii committed:
add gps data to datagenerator

MCLiii committed:
Ammend pull run script again because network=host doesn't work on windows

MCLiii committed:
Oh god I hate windows

MCLiii commented:
Should be ready to merge, any comments?

MCLiii committed:
touch ups

MCLiii committed:
remove old code and minor bug fix

jackwherry commented on Backend/components/graph_api.py:
```diff
@@ -0,0 +1,55 @@
+
+@router.get("/graph")
+async def get_graph(data: str, start_time: int, end_time: int):
+    """pass in a comma separated list of data that the frontend requests start_time end_time must be in unix millisec"""
```
```suggestion
    """pass in a comma separated list of data that the frontend requests. start_time and end_time must be in unix milliseconds"""
```

jackwherry commented on Backend/components/graph_api.py:
```diff
@@ -0,0 +1,55 @@
+    # sort the data into smaller group by aggregation methods
+    group = {'avg': [], 'first':[], 'last':[], 'max': [], 'min':[], 'no-aggr':[], 'sum': []}
+    for d in data:
+        # TODO change this to actual aggregation specified in data format
```
did you want to finish these todos before merging this or are you ok leaving them?

jackwherry replied:
If they get left we should probably make an issue tracking it so that we don't forget

MCLiii replied:
Good point, those are waiting on update to data-format, I'm making an issue for them.

jackwherry commented on Backend/config.py:
```diff
@@ -0,0 +1,17 @@
+REDIS_DB = 0
+
+CAR_IP = '192.168.1.15'
+LOCAL_IP = 'localhost'
```
Is there a reason why this is localhost but it's 127.0.0.1 for redis?

MCLiii replied:
Nope

jackwherry commented on Backend/core/comms.py:
```diff
@@ -0,0 +1,181 @@
+                solar_car_connection['tcp'] = True
+                print('connected')
+            except Exception:
+                time.sleep(3)   # macOS is the superior system, it teaches you an important lesson of having patience
```
Is this reliable enough? I'd imagine there's another way to avoid this issue with async or forcing the connection to be blocking, but I haven't experimented so this solution is just fine if it works reliably

MCLiii replied:
Yeah, this is a temporary fix, I believe it has something to do with the macOS implementation taking up too much thread time and blocking other threads from executing. we should definitely look into better practices for this, I'll make this one an issue as well.

MCLiii replied:
... actually, since we are switching to udp soon and since udp is connectionless, it should eliminate thread time on retrying. We'll see how udp preforms and decide if we need performance improvement.

jackwherry commented on Backend/core/comms.py:
```diff
@@ -0,0 +1,181 @@
+        :param tmp_source: Name of tmp data source, put comm channel name here e.g. tcp, lte
+        """
+        header = b'<bsr>'
+        footer = b'</bsr>'
```
Does it make any sense to move this out to a config file variable or is it not expected to ever change again?

MCLiii replied:
I would assume we won't have a name change soon...

jackwherry commented on Backend/core/comms.py:
```diff
@@ -0,0 +1,181 @@
+        # If the remaining data is longer than the expected packet length,
+        # there might be an incomplete packet, so log a warning.
+        if len(self.__tmp_data[tmp_source]) >= byte_length:
+            print("ERROR: Incomplete or malformed packet ------------------------------------")
```
Should this message be changed to say "warning" instead of "error" since it doesn't stop the program and doesn't really require the user to do anything?

MCLiii replied:
True

jackwherry commented on Backend/core/comms.py:
```diff
@@ -0,0 +1,181 @@
+    # Start two live comm channels
+    vps_thread = threading.Thread(target=lambda : asyncio.run(telemetry.remote_db_fetch(config.VPS_URL)))
+    vps_thread.start()
+    #tcp.listen_tcp(config.LOCAL_IP if len(sys.argv) > 1 and sys.argv[1]=='dev' else config.CAR_IP, config.DATA_PORT)
```
Should probably remove commented out code unless there's a good reason for it

jackwherry reviewed (approved):
Couple minor comments but looks good

MCLiii committed:
update readme

MCLiii committed:
Address PR suggestions

jrvollmer commented on Backend/core/core_api.py:
```diff
@@ -0,0 +1,48 @@
+    if comms.solar_car_connection['tcp'] or comms.solar_car_connection['lte']:
+        latest_data = comms.frontend_data
+        latest_data['solar_car_connection'] = True
+        latest_data['timestamps'] = f'{latest_data["tstamp_hr"]:02d}:{latest_data["tstamp_mn"]:02d}:' \
+                                    f'{latest_data["tstamp_sc"]:02d}.{latest_data["tstamp_ms"]}'
```
Shouldn't we be using the unix ts?

MCLiii replied:
The frontend is still using 24 hour format

jrvollmer commented on Backend/core/core_api.py:
Could we split up data recording and single values api endpoints?

MCLiii replied:
Yes

jrvollmer commented on Backend/Data/gps_dataset/dataset1.csv:
2 things:
1. gps_dataset/*.csv/gpx/json/whatever extensions will be generated output should be in the .gitignore (you'll have to do some git magic to make this change in the comimt history before any of the files were added/committed. Otherwise, it won;t have any affect
2. You can add a .gitkeep or README.md (probably the latter so that you can have a quick explanation of the purpose of the directory in the repo) to be able to commit the directory.

jrvollmer commented on .gitignore:
```diff
@@ -26,6 +26,9 @@
 
+# Python
+/Backend/*.egg-info
+
```
Maybe move this into the Backend .gitignore

jrvollmer commented on Backend/setup.py:
```diff
@@ -0,0 +1,13 @@
+
+setup(
+    name='chase-car-dashboard-backend',
+    version='5.1.0',
```
Where was version 5.0.0?

jrvollmer commented on DataGenerator/src/server.js:
```diff
@@ -1,4 +1,5 @@
 import DATA_FORMAT from "../../Backend/Data/sc1-data-format/format.json";
+import GPS_DATA from "../../Backend/Data/gps_dataset/dataset1.json";
```
What are we doing with this? This should just be generating random data

jrvollmer commented on Frontend/package-lock.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "chase-car-dashboard-frontend",
-  "version": "4.2.0",
+  "version": "4.3.0",
```
New api endpoints with the backend -> major version update

jrvollmer commented on Frontend/src/Components/Dashboard/DataRecordingControl.js:
```diff
@@ -66,139 +66,27 @@ export default function DataRecordingControl(props) {
     // ------------------------------------------ Data recording controls --------------------------------------------
 
-    const { isOpen, onOpen, onClose } = useDisclosure();
+    // TODO: consider making isProcessing boolean to control when to display a loading spinner
```
Can we address TODOs before merging?

MCLiii replied:
@jackwherry

jackwherry replied:
Removed from source code and moved to an issue on Jira (pretty low priority for now though): https://badgerloop.atlassian.net/browse/SW1-24?atlOrigin=eyJpIjoiODY2NzkxMThmZWY3NDFhOGE2ODk5YjliMTA2YTM1NTUiLCJwIjoiaiJ9

jrvollmer commented on package.json:
```diff
@@ -1,17 +1,15 @@
 {
   "name": "chase-car-dashboard",
-  "version": "4.2.0",
+  "version": "5.0.0",
```
I agree with this version change, but it has to match Backend's version, so can you update Backend/setup.py's version to be 5.0.0?

MCLiii committed:
move data recording out of core_api.py

MCLiii committed:
Add additional checks to byte array length

MCLiii committed:
rearrange git ignore

MCLiii committed:
move dataset folder

MCLiii committed:
change datagenerator path

jackwherry committed:
migrate data recording control TODO to SW1-1/SW1-24

MCLiii committed:
set version number

MCLiii commented:
@jrvollmer 👀👀👀👀👀👀👀

jrvollmer commented on Backend/framework/router.py:
This is more of a comment, but not blocking. Why do we have single-values in `core/` and graph_api/recorded_data in `components/`? I feel like they should all be in `components/`. Maybe `components/` should also be renamed `routes/`?

MCLiii replied:
Single value directly shares data with comms, but other apis are using data gathered from the database thus in components. I think components are more descriptive of what the things in the folders are than route.

jrvollmer commented:
Just updated a version number in the frontend and reverted the 5.1.0 version in setup.py

jrvollmer reviewed (approved):
Here ya go @MCLiii. LGTM!!!!!!!!!!!!!!
